plugins/volume/org.xen.xapi.storage.zfs-ng/sr.py

```diff
@@ -0,0 +1,226 @@
+
+import importlib
+
+ZPOOL_BIN = 'zpool'
+ZFS_BIN = 'zfs'
+
+@util.decorate_all_routines(util.log_exceptions_in_function)
+class Implementation(xapi.storage.api.v5.volume.SR_skeleton):
+    def probe(self, dbg, configuration):
+        return {
+            'srs': [],
+            'uris': []
+        }
+
+    def attach(self, dbg, configuration):
+        uri = configuration['mountpoint']
+        log.debug('{}: SR.attach: config={}, uri={}'.format(
+            dbg, configuration, uri))
+
+        sr = urlparse.urlparse(uri).path
+        return sr
+
+    def create(self, dbg, sr_uuid, configuration, name, description):
+        log.debug('{}: SR.create: config={}, sr_uuid={}'.format(
+            dbg, configuration, sr_uuid))
+
+        if 'devices' not in configuration:
+            log.error('devices parameter is missed')
+            raise
+        devs = configuration['devices'].split(',')
+
+        mountpoint = '/' + name
+        if 'mountpoint' in configuration:
+            mountpoint = configuration['mountpoint']
+
+        compression = False
+        if 'compression' in configuration:
+            if configuration['compression'] in [
+                    'true', 't', 'on', '1', 'yes']:
+                compression = True
+
+        cmd = [
+            ZPOOL_BIN, 'create', '-f',
+            name, '-m', mountpoint
+        ]
+
+        if 'mode' in configuration:
+            if configuration['mode'] not in [
+                    'N', 'M', 'R']:
+                log.error('mode can only be N(default), M(mirror) or R(raidz)')
+                raise
+            if configuration['mode'] in ['M']:
+                if len(devs) < 2:
+                    log.error('mirror mode requires at least two devices')
+                    raise
+                cmd.append('mirror')
+            if configuration['mode'] in ['R']:
+                if len(devs) < 2:
+                    log.error('raidz mode requires at least two devices')
+                    raise
+                cmd.append('raidz')
+
+        cmd.extend(devs)
+
+        try:
+            call(dbg, cmd)
+        except:
+            log.error('error creating the pool')
+            raise
+
+        if compression:
+            cmd = [
+                ZFS_BIN, 'set', 'compression=on', name
+            ]
+            call(dbg, cmd)
+
+        log.debug('{}: SR.create: sr={}'.format(dbg, mountpoint))
+
+        importlib.import_module('zfs-ng').Callbacks().create_database(mountpoint)
+
+        meta = {
+            'name': name,
+            'description': description,
+            'uri': mountpoint,
+            'mountpoint': mountpoint,
+            'unique_id': sr_uuid,
+            'read_caching': False,
+            'keys': {}
+        }
+        util.update_sr_metadata(dbg, 'file://' + mountpoint, meta)
+
+        return configuration
+
+    def destroy(self, dbg, sr):
+        name = os.path.basename(sr)
+        cmd = [
+            ZPOOL_BIN, 'destroy',
```

Comment:
Use ZFSUtil?

Reply:
ZFSUtil.py:destroy() uses ZFS_BIN instead of ZFS_POOL. That's why I do not use ZFSUtil for ZPOOL_BIN commands.

Reply:
I added destroy_pool() in ZFSUtil.